keps/973-workload-priority/README.md

```diff
@@ -0,0 +1,225 @@
+
+#### Story 1
+
+In an organization, admins want to set a lower priority for development workloads and a higher priority for production workloads.
+In such cases, they create two `WorkloadPriorityClass` and apply each one to the respective workloads.
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha1
+kind: WorkloadPriorityClass
+metadata:
+  name: dev-workload-class
+value: 100
+description: "Dev's priority class"
+---
+apiVersion: kueue.x-k8s.io/v1alpha1
+kind: WorkloadPriorityClass
+metadata:
+  name: prod-workload-class
+value: 10000
+description: "Prod's priority class"
+---
+# dev-workload-a.yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: dev-job-a
+spec:
+  priorityClassName: dev-workload-class
+  queueName: team-a-queue
+  podSets:
+  - count: 3
+    name: main
+    template:
+      spec:
+        containers:
+        - image: gcr.io/k8s-staging-perf-tests/sleep:latest
+---        
+# dev-workload-b.yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: dev-job-b
+spec:
+  priorityClassName: dev-workload-class
+  queueName: team-b-queue
+  podSets:
+  - count: 3
+    name: main
+    template:
+      spec:
+        containers:
+        - image: gcr.io/k8s-staging-perf-tests/sleep:latest
+# prod-workload-a.yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: prod-job-b
+spec:
+  priorityClassName: prod-workload-class
+  queueName: team-b-queue
+  podSets:
+  - count: 3
+    name: main
+    template:
+      spec:
+        containers:
+        - image: gcr.io/k8s-staging-perf-tests/sleep:latest        
+```
+
+
+
+### Risks and Mitigations
+
+It's possible that the pod's priority conflicts with the workload's priority.
+For example, a high-priority job with low-priority pods may never run to completion because it may always be preempted.
+We should document the risks of pod preemption to uses.
+We can also point users to create `PriorityClass` for their pods that are [non-preempting](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class).
+
+## Design Details
+
+### Kueue WorkloadPriorityClass API
+
+We introduce the `WorkloadPriorityClass` API.
+
+```golang
+type WorkloadPriorityClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Value int32   `json:"value,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+```
+
+### How to use WorkloadPriorityClass on Workload
+
+The `WorkloadPriorityClass` is defined using `PriorityClassName` field.
+
+```yaml
+# sample-priority-class.yaml
+apiVersion: kueue.x-k8s.io/v1alpha1
+kind: WorkloadPriorityClass
+metadata:
+  name: sample-priority
+value: 10000
+description: "Sample priority"
+---
+# sample-workload.yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: sample-job
+spec:
+  priorityClassName: sample-priority
```

Comment:
there should already be a priorityValue. We should populate it from the Workload reconciler if it's not populated.

Reply:
Thanks. Updated